### PR TITLE
JAXB - Ignore @XmlTransient fields/methods when marking hierarchy for reflection

### PR DESCRIPTION
**Pull request by gsmet**
This is a follow up of https://github.com/quarkusio/quarkus/pull/37643 .

### CODE REVIEW

**Review by geoand — APPROVED** [review]

**Review by gastaldi — APPROVED** [review]

**Comment by gsmet** [comment]
This will need more work.

**Comment by rysurd** [comment]
> This will need more work.

If I can provide any help let me know, as I have worked on the fix that cause this issue

**Comment by ppalaga** [comment]
This does not fix Quarkus CXF. I'd vote for going with https://github.com/quarkusio/quarkus/pull/38221

**Comment by gsmet** [comment]
Could you provide specifics about the quarkus-cxf issue, especially how I can reproduce it?

**Comment by ppalaga** [comment]
> Could you provide specifics about the quarkus-cxf issue

Using ReflectiveHierarchyBuildItem is just too risky. It pulls too much stuff in, the most of which does not help to fix https://github.com/quarkusio/quarkus/issues/36479 at all. To fix it is enough to climb up the superclasses and that's all we should do. 

> especially how I can reproduce it?

Easy: build [QCXF](https://github.com/quarkiverse/quarkus-cxf/tree/main/) with Quarkus 999-SNAPSHOT 

```
mvnd clean install -DskipTests -Dquarkus.build.skip

# and run the integration tests in native mode:

cd integration-tests
mvnd clean verify -Pnative -Dquarkus.native.container-build -Dtest=foo -Dsurefire.failIfNoSpecifiedTests=false -1
```

**Comment by gsmet** [comment]
@ppalaga I think you are being a bit too assertive here.

Going through superclasses does not fix everything. And we introduced `ReflectiveHierarchyBuildItem` precisely for the exact same use case for JSON. So using it for JAXB does actually makes sense.

Now, I see two problems: we did not handle things carefully enough for ignoring fields and methods (this is in progress). And even with that, we still have things leaking but I wouldn't say it always comes from this new feature.

Typically I would expect this getter: https://github.com/apache/camel/blob/main/core/camel-core-model/src/main/java/org/apache/camel/model/OptionalIdentifiedDefinition.java#L49 to be transient and not the setter only. And I suppose there are other things problematic in Camel.

As for CXF, I was able to get at least several modules to build in native with my new patch but... I have no idea which modules are failing for you given the level of details you gave me (I know how to build something, what I want to know is what exactly is the problem and which modules are failing).

**Comment by gsmet** [comment]
My current proposal is to revert this in 3.7.0.CR1 but keep it in main and make progress on both sides during the 3.9 cycle which will be extra long.

Unfortunately it already slipped in 3.6.6 so don't update your projects to it.

Would it work for you @jamesnetherton @ppalaga ?

**Comment by jamesnetherton** [comment]
>  revert this in 3.7.0.CR1 but keep it in main and make progress on both sides during the 3.9 cycle

That would be my preference assuming it wont leak into 3.8.x.

**Comment by gsmet** [comment]
@jamesnetherton 3.8 will be branched from 3.7 so it won't.

But in any case, I want to test something in the release process tomorrow and I need everything to be in `main` so I'm reverting it in `main` and will prepare another PR to reintroduce it + the fix and make sure you guys test it and don't have weird failures.

**Comment by gsmet** [comment]
See https://github.com/quarkusio/quarkus/pull/38224

**Comment by ppalaga** [comment]
> My current proposal is to revert this in 3.7.0.CR1

Thanks a lot!

> but keep it in main

> And we introduced ReflectiveHierarchyBuildItem precisely for the exact same use case for JSON. So using it for JAXB does actually make sense.

The classes I saw registered for CXF looked like ancillary stuff put in when generating the JAXB classes from Schemas and WSDLs. I guess it will be a long run to make it all work with ReflectiveHierarchyBuildItem by either excluding manually or fixing those generators. https://github.com/quarkusio/quarkus/issues/36479 complained about missing fields from a superclass and that's also what the provided test reproduces. I did not have the impression that the issue is more general (sorry if I missed something). If that's the case anyway, there should be tests for situations that show the necessity to use the ReflectiveHierarchyBuildItem.

**Comment by gsmet** [comment]
@ppalaga if you can come up with pointers to classes that are handled incorrectly, that would be helpful. Ideally with the name of the problematic field/method. I think there are probably not that many corner cases to handle (but I'm not sure we have all the info to do it properly, we will see).

**Comment by jamesnetherton** [comment]
I reran the previously failing tests in CQ and they all passed. Also tested a few other itest modules that use JAXB and they passed too.

File: integration-tests/jaxb/src/main/java/io/quarkus/it/jaxb/BookIBANField.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package io.quarkus.it.jaxb;
+
+import jakarta.xml.bind.annotation.XmlAccessType;
+import jakarta.xml.bind.annotation.XmlAccessorType;
+import jakarta.xml.bind.annotation.XmlElement;
+import jakarta.xml.bind.annotation.XmlTransient;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlTransient
```
Comment (ppalaga) [inline integration-tests/jaxb/src/main/java/io/quarkus/it/jaxb/BookIBANField.java:9]: Adding `@XmlTransient` here excludes BookIBANField from reflection registration, no? In that way we stop reproducing the original use case of https://github.com/quarkusio/quarkus/issues/36479 . Shouln't we better introduce a new class hierarchy for testing the exclusion of `@XmlTransient`?

**Comment by ppalaga** [comment]
> @ppalaga if you can come up with pointers to classes that are handled incorrectly, that would be helpful. Ideally with the name of the problematic field/method. I think there are probably not that many corner cases to handle (but I'm not sure we have all the info to do it properly, we will see).

All tests failing with the original change can be seen in this PR: https://github.com/quarkiverse/quarkus-cxf/pull/1175 
It is basically all that have a `native` profile under https://github.com/quarkiverse/quarkus-cxf/tree/main/integration-tests
I unfortunately won't be able to try your newest proposal before Thursday afternoon.

**Comment by gsmet** [comment]
There’s no hurry. I reverted the patch for 3.7. This will be for main only. 

I didn’t add tests for now. I wanted to make sure the approach had value before spending time on tests.

File: extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java [hunk 2]
```diff
@@ -421,4 +447,92 @@ private void addResourceBundle(BuildProducer<NativeImageResourceBundleBuildItem>
...
+                return true;
+        }
+    }
+
+    private static MethodInfo getGetterSetterCounterPart(MethodInfo methodInfo) {
+        if (!methodInfo.name().startsWith("get") || methodInfo.parametersCount() > 0) {
+            return null;
+        }
+
+        return methodInfo.declaringClass().method(methodInfo.name().replaceFirst("get", "set"), methodInfo.returnType());
```
[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:522]: We only consider `get` for reflective hierarchy logic as `is` will return a boolean and we don't need to register it.

[the remark this answers is not in the document]
Comment (ppalaga) [reply extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java:522]: Thanks, sounds legit. Is it the same for `java.lang.Boolean`?

**Review by ppalaga — APPROVED** [review]
Sorry for the delay, @gsmet. I have just run some audits on Quarkus CXF to see the impact and I can confirm that this fix is not adding any new reflection registrations in Quarkus CXF tests. That's a good sign, thanks for you great work!

**Comment by gsmet** [comment]
Thanks for the confirmation @ppalaga ! I'll run CI and merge once it's ready.

**Comment by rysurd** [comment]
Hello ! This fix will be deployed in the 3.9 cycle if I understand well ? Or is for another version ?

**Comment by gsmet** [comment]
Yes, except if there is a strong push for it, it's going to be in 3.9 as it's a bit too involved to get backported.
